gio-subclass/src/application.rs

```diff
@@ -0,0 +1,655 @@
+use gio;
+use gio_ffi;
+use glib;
+use glib::translate::*;
+use glib::IsA;
+use glib_ffi;
+use gobject_ffi;
+use libc;
+use std::mem;
+use std::ptr;
+use std::ops::Deref;
+use std::fmt;
+use std::convert;
+
+use gobject_subclass::anyimpl::*;
+use gobject_subclass::object::*;
+
+pub struct ArgumentList{
+    pub(crate) ptr: *mut *mut *mut libc::c_char,
+    items: Vec<String>
+}
+
+impl ArgumentList{
+
+    pub(crate) fn new(arguments: *mut *mut *mut libc::c_char) -> Self{
+        Self{
+            ptr: arguments,
+            items: unsafe{
+                FromGlibPtrContainer::from_glib_none(ptr::read(arguments))
+            }
+        }
+    }
+
+    // remove the item at index `idx` and shift the raw array
+    pub fn remove(&mut self, idx: usize){
+
+        unsafe{
+            let n_args = glib_ffi::g_strv_length(*self.ptr);
+            assert!((n_args as usize) == self.items.len());
+            assert!((idx as u32) < n_args);
+
+            self.items.remove(idx);
+
+            glib_ffi::g_free(((*self.ptr).offset(idx as isize)) as *mut libc::c_void);
+
+            for i in (idx as u32)..n_args-1 {
+                ptr::write((*self.ptr).offset(i as isize), *(*self.ptr).offset((i+1) as isize))
+            }
+            ptr::write((*self.ptr).offset((n_args-1) as isize), ptr::null_mut());
+        }
+    }
+}
+
+impl Deref for ArgumentList{
+    type Target = [String];
+
+    fn deref(&self) ->&Self::Target{
+        self.items.as_slice()
+    }
+}
+
+impl fmt::Debug for ArgumentList{
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error>{
+        self.items.fmt(formatter)
+    }
+}
+
+impl convert::Into<Vec<String>> for ArgumentList{
+    fn into(self) -> Vec<String>{
+        self.items.clone()
+    }
+}
+
+
+pub trait ApplicationImpl<T: ApplicationBase>: ObjectImpl<T> + AnyImpl + 'static {
+    fn startup(&self, application: &T) {
+        application.parent_startup();
+    }
+
+    fn activate(&self, application: &T) {
+        application.parent_activate();
+    }
+
+    fn open(&self, application: &T, files: &Vec<gio::File>, hint: &String) {
```

Comment:
As this array is not mutable in any way, maybe just pass a `&[gio::File]`. That's more generic